napari/_qt/dialogs/qt_package_installer.py

```diff
@@ -0,0 +1,419 @@
+        super().__init__(parent)
+        self._queue: Deque[AbstractInstallerTask] = Deque()
+        self._output_widget = None
+
+        self.setProcessChannelMode(QProcess.MergedChannels)
+        self.readyReadStandardOutput.connect(self._on_stdout_ready)
+        self.readyReadStandardError.connect(self._on_stderr_ready)
+
+        self.finished.connect(self._on_process_finished)
+        self.errorOccurred.connect(self._on_error_occurred)
+
+    # -------------------------- Public API ------------------------------
+    def install(
+        self,
+        task_handler: TaskHandler,
+        pkgs: Sequence[str],
+        *,
+        prefix: Optional[str] = None,
+        origins: Sequence[str] = (),
+        **kwargs,
+    ) -> JobId:
+        """Install packages in `pkgs` into `prefix` using `task_handler` tool with additional
+        `origins` as source for `pkgs`.
+
+        Parameters
+        ----------
+        task_handler : str
+            Which type of installation task_handler to use: pip or conda
+        pkgs : Sequence[str]
+            List of packages to install.
+        prefix : Optional[str], optional
+            Optional prefix to install packages into.
+        origins : Optional[Sequence[str]], optional
+            Additional sources for packages to be downloaded from.
+
+        Returns
+        -------
+        JobId : int
+            ID that can be used to cancel the process.
+        """
+        item = self._build_queue_item(
+            task_handler=task_handler,
+            action="install",
+            pkgs=pkgs,
+            prefix=prefix,
+            origins=origins,
+            **kwargs,
+        )
+        return self._queue_item(item)
+
+    def uninstall(
+        self,
+        task_handler: TaskHandler,
+        pkgs: Sequence[str],
+        *,
+        prefix: Optional[str] = None,
+        **kwargs,
+    ) -> JobId:
+        """Uninstall packages in `pkgs` from `prefix` using `task_handler` tool.
+
+        Parameters
+        ----------
+        task_handler : str
+            Which type of installation task_handler to use: pip or conda
+        pkgs : Sequence[str]
+            List of packages to uninstall.
+        prefix : Optional[str], optional
+            Optional prefix from which to uninstall packages.
+
+        Returns
+        -------
+        JobId : int
+            ID that can be used to cancel the process.
+        """
+        item = self._build_queue_item(
+            task_handler=task_handler,
+            action="uninstall",
+            pkgs=pkgs,
+            prefix=prefix,
+            **kwargs,
+        )
+        return self._queue_item(item)
+
+    def cancel(self, job_id: Optional[JobId] = None):
+        """Cancel `job_id` if it is running.
+
+        Parameters
+        ----------
+        job_id : Optional[JobId], optional
+            Job ID to cancel.  If not provided, cancel all jobs.
+        """
+
+        def end_process():
```

Comment:
I would rather we do not use nested funcs. Could we move this nested func outside as a private static method inside the class... or as a private func outside the class (plus a docstring :) )?